### PR TITLE
Split barriers with flush and data

### PR DESCRIPTION
**Pull request by SoujanyaPonnapalli**

### CODE REVIEW

File: code/permuter/Permuter.cpp [hunk 1]
```diff
@@ -96,6 +107,34 @@ void Permuter::InitDataVector(vector<disk_write> *data) {
...
+      	if(data->at(index).has_flush_flag()) {
+      		flag_half.set_flush_flag();
+      		data_half.clear_flush_flag();
+      	}
+      	if(data->at(index).has_flush_seq_flag()) {
+      		flag_half.set_flush_seq_flag();
+      		data_half.clear_flush_seq_flag();
+      	}
+      	
+        epoch_op curr_op = {index, flag_half};
```
Comment (vijay03) [inline code/permuter/Permuter.cpp]: Spacing not correct?

File: code/permuter/Permuter.cpp [hunk 2]
```diff
@@ -96,6 +107,34 @@ void Permuter::InitDataVector(vector<disk_write> *data) {
...
+      	current_epoch.ops.push_back(curr_op);
+      	current_epoch.num_meta += flag_half.is_meta();
+      	epochs_.push_back(current_epoch);
+      	prev_epoch_flush_op = true;
+        current_epoch.has_barrier = true;
+        ++index;
```
Comment (vijay03) [inline code/permuter/Permuter.cpp]: Similar spacing issue.

File: code/permuter/Permuter.cpp [hunk 3]
```diff
@@ -60,6 +63,14 @@ void Permuter::InitDataVector(vector<disk_write> *data) {
     // Get all ops in this epoch and add them to either the sync_op or async_op
     // lists.
     while (index < data->size() && !(data->at(index)).is_barrier_write()) {
+      
+      if (prev_epoch_flush_op == true) {
```
Comment (ashmrtn) [inline code/permuter/Permuter.cpp]: nit: indentation. 2 space tabs please :)

File: code/permuter/Permuter.cpp [hunk 4]
```diff
@@ -96,6 +107,34 @@ void Permuter::InitDataVector(vector<disk_write> *data) {
     // to the special spot in the epoch, otherwise just push the current epoch
     // onto the list and move to the next segment of the log.
     if (index < data->size() && (data->at(index)).is_barrier_write()) {
+      
+      // Check if the op at the current index has a flush flag with data. It it has, then divide
+      // it into two halves and make the data available only in the starting of the next epoch.
```
Comment (ashmrtn) [inline code/permuter/Permuter.cpp]: nit: "starting" -> "start"

File: Soujanya [hunk 5]
```diff
@@ -0,0 +1,512 @@
+[33mcommit 893f63e5858c92d9078ff2519f55e99366351c9f[m
```
Comment (ashmrtn) [inline Soujanya]: Remove file please

**Comment by vijay03** [comment]
@ashmrtn : I believe @SoujanyaPonnapalli has done all requested changes right? Could you review and approve?

**Comment by ashmrtn** [comment]
Oh, right, sorry! I'll check this evening when I get time!

File: code/permuter/Permuter.cpp [hunk 6]
```diff
@@ -51,6 +51,9 @@ bool BioVectorEqual::operator() (const std::vector<unsigned int>& a,
 void Permuter::InitDataVector(vector<disk_write> *data) {
   epochs_.clear();
   unsigned int index = 0;
+  bool prev_epoch_flush_op = false;
+  disk_write data_half;
+  // disk_write data_flush_part1, data_write_part2;
```
Comment (ashmrtn) [inline code/permuter/Permuter.cpp]: If this is unused could you remove it?

File: code/permuter/Permuter.cpp [hunk 7]
```diff
@@ -60,6 +63,14 @@ void Permuter::InitDataVector(vector<disk_write> *data) {
     // Get all ops in this epoch and add them to either the sync_op or async_op
     // lists.
     while (index < data->size() && !(data->at(index)).is_barrier_write()) {
+      
+      if (prev_epoch_flush_op == true) {
+      	epoch_op curr_op = {index-1, data_half};
```
Comment (ashmrtn) [inline code/permuter/Permuter.cpp]: This still looks like one space?

File: code/permuter/Permuter.cpp [hunk 8]
```diff
@@ -96,6 +107,35 @@ void Permuter::InitDataVector(vector<disk_write> *data) {
...
+      
+      // Check if the op at the current index has a flush flag with data. It it has, then divide
+      // it into two halves and make the data available only in the start of the next epoch.
+      // If the op has a FUA flag, then it gets normally added into the current epoch
+      if ((data->at(index).has_flush_flag() || data->at(index).has_flush_seq_flag()) && data->at(index).has_write_flag() && (!data->at(index).has_FUA_flag())) {
+        
+        disk_write flag_half;
+        data_half = data->at(index);
+        
+        if(data->at(index).has_flush_flag()) {
```
Comment (ashmrtn) [inline code/permuter/Permuter.cpp]: Space between opening parentheses for the if statement

File: code/permuter/Permuter.cpp [hunk 9]
```diff
@@ -96,6 +107,35 @@ void Permuter::InitDataVector(vector<disk_write> *data) {
...
+          flag_half.set_flush_flag();
+          data_half.clear_flush_flag();
+        }
+        
+        if(data->at(index).has_flush_seq_flag()) {
```
Comment (ashmrtn) [inline code/permuter/Permuter.cpp]: Same as above

**Pull request merged**